design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,299 @@
+ordering of updates to a specific key are maintained across the repartition.
+
+## Test plan
+
+Aside from the usual unit tests etc, the QTT suit of tests will be extended to cover the different
+key formats. Tests will be added to cover the new syntax and configuration combinations and 
+permutations. Existing tests covering aggregations, re-partitions and joins will be extended to 
+include variants with different key formats. 
+
+## LOEs and Delivery Milestones
+
+The KLIP will be broken down into the following deliverables:
+
+1. **Basic JSON support**: Support for the `JSON` key format, without:
+    * schema registry integration
+    * making KAFKA_TOPIC property optional
+    * Automatic repartitioning of streams and tables for joins where key formats do not match: such
+      joins will result in an error initially.
+  
+    Included in this milestone:
+    
+    * Addition of a new optional `KEY_FORMAT` property in the `WITH` clause, to set the key format.
+    * Addition of a new optional `FORMAT` property in the `WITH` clause, to set both the key & value formats.
+    * Addition of new server configuration to provide defaults for key and value formats.
+    * Removal of requirement for `VALUE_FORMAT` property in the `WITH` clause of CT/CS statements, where
+      the server configuration provides a default.
+    * Support for additional key column data types, as JSON supports them:
+        * `DECIMAL`
+        * `BOOLEAN`
+    * Full support of these key formats for all supported SQL syntax  
+1. **Auto-repartitioning on key format mismatch**. Adds support for automatic repartitioning of 
+   streams and tables for joins where key formats do not match.
+1. **Schema Registry support**: Adds support for reading and writing schemas to and from the schema
+   registry.
+1. **JSON_SR support** Adds support for the `AVRO` key format, inc. schema registry integration.
+1. **Avro support** Adds support for the `AVRO` key format, inc. schema registry integration.
+1. **Delimited support**: Adds support for the `DELIMITED` key format.
+1. **Optional KAFKA_TOPIC property**: makes the `KAFKA_TOPIC` property optional everywhere and add
+   configuration to control case and case sensitivity of topic names.
+1. **Blog post**: write a blog post about the new features. (Potentially more than once if 
+  work span multiple releases).
+   
+LOE TBD once scope and design confirmed.
+
+## Documentation Updates
+
+New server config and new `CREATE` properties will be added to main docs site.
+
+There are no incompatible changes within the proposal, so no demos and examples _must_ change.
+However, it probably pays to update some to highlight the new features. We propose updating the 
+Kafka micro site examples to leverage the new functionality, as these have automated testing.  
+It may be worth changing the ksqlDB quickstart too - TBD, as this will require extending DataGen 
```

Comment:
Which quickstart does this refer to? Neither the on-prem nor cloud quickstart uses ksql-datagen.

Reply:
Good to know!

Reply:
Updated.